### PR TITLE
Add `cupy.polyder` API.

### PR DESCRIPTION
**Pull request by khushi-411**
Hi,
The PR follows a part of #6078. It aims to:
- Implement `cupy.polyder` API. Test added.
- Use NumPy Style docstring in the file as suggested in #5437.

Looking forward to hearing from you. Thanks :)

### CODE REVIEW

File: cupy/lib/_routines_poly.py [hunk 1]
```diff
@@ -104,6 +125,43 @@ def polymul(a1, a2):
...
+    Parameters
+    ----------
+    p : poly1d or sequence
+        Polynomial to differentiate
+    m : int, optional
+        Order of differentiation. By default, 1
+
+    Returns
+    -------
+    der : poly1d
```
Comment (asi1024) [inline cupy/lib/_routines_poly.py:140]:
```suggestion
    der : cupy.ndarray or cupy.poly1d
```

File: cupy/lib/_routines_poly.py [hunk 2]
```diff
@@ -124,41 +182,51 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
...
+    cov : bool or str, optional
+        If given, returns the coefficients
+        along with the covariance matrix.
+
+    Returns
+    -------
+    p : cupy.ndarray of shape (deg + 1,) or (deg + 1, K)
+        Polynomial coefficients from highest to lowest degree
+    residuals, rank, singular_values, rcond :
+    cupy.ndarray, int, cupy.ndarray, float
```
Comment (asi1024) [inline cupy/lib/_routines_poly.py:213]: It seems that the documentation is not generated expectedly.
https://cupy--6469.org.readthedocs.build/en/6469/reference/generated/cupy.polyfit.html

Comment (khushi-411) [reply cupy/lib/_routines_poly.py]: Yeah, thanks for pointing it out. I've updated it :)

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 3]
```diff
@@ -843,6 +843,48 @@ def test_roots_poly1d(self, xp, dtype):
         return xp.roots(xp.poly1d(a))
 
 
+@testing.gpu
+class TestPolyDer:
```
[the remark this answers is not in the document]
Comment (khushi-411) [reply tests/cupy_tests/lib_tests/test_polynomial.py:847]: Done

File: cupy/lib/_routines_poly.py [hunk 4]
```diff
@@ -104,6 +125,43 @@ def polymul(a1, a2):
...
+        raise ValueError("Order of derivative must be positive.")
+
+    truepoly = isinstance(p, cupy.poly1d)
+    p = cupy.asarray(p)
+    n = len(p) - 1
+    y = p[:-1] * cupy.arange(n, 0, -1)
+    if m == 0:
+        val = p
+    else:
+        val = polyder(y, m - 1)
```
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/lib/_routines_poly.py:159]: How about this comment?

**Comment by asi1024** [comment]
@khushi-411 Any updates?

**Comment by khushi-411** [comment]
Hi, @asi1024!
Apologies for the delay. I'll update the PR soon (surely by today or tomorrow). Thanks for the ping. :)

**Comment by khushi-411** [comment]
Hi, @asi1024!
Could you please take a look at the following code snippet?
```python
_polyder_kernel = _core.ElementwiseKernel(
    'T p, int64 m',
    'T z',
    '''
    int n = sizeof(p)/sizeof(p[0]) - 1;
    T y = p[:-1] * cupy.arange(n, 0, -1);
    if (m == 0) {
        z = p;
    } else {
        z = _polyder_kernel(y, m - 1);
    }
    ''',
    '_polyder_kernel')
```
The tests are failing if using the kernel. Here's the error [link](https://pastebin.com/sFqzjTkB). Please let me know some tips to correct it. Thanks!

**Comment by asi1024** [comment]
@khushi-411 ElementwiseKernel body string parameter is embedded as CUDA program in a CUDA __global__ function, so cupy ufuncs cannot be used in this string.

File: cupy/lib/_routines_poly.py [hunk 5]
```diff
@@ -147,6 +168,52 @@ def polymul(a1, a2):
...
+_polyder_kernel = cupy._core.ElementwiseKernel(
+    'raw T p, int64 m',
+    'T z',
+    '''
+    T y = p[:-1] * p[_ind.size() - i - 1][1:];
+    if (m == 0) {
+        z = p;
+    } else {
+        z = polyder(y, m - 1);
+    }
```
Comment (khushi-411) [inline cupy/lib/_routines_poly.py:180]: There is a problem with indexing, I think. Any hints about the same? Thanks!

Comment (asi1024) [reply cupy/lib/_routines_poly.py:180]: This string is embedded as a body of the CUDA `__global__` function, so Python-style array slicing is not available here.

File: tests/cupy_tests/lib_tests/test_polynomial.py [hunk 6]
```diff
@@ -904,6 +904,54 @@ def test_roots_poly1d(self, xp, dtype):
...
```
Comment (asi1024) [inline tests/cupy_tests/lib_tests/test_polynomial.py:908]: Could you add tests also for ndarray type inputs?

**Comment by asi1024** [comment]
@khushi-411 Any updates?

**Comment by asi1024** [comment]
Closing the PR as it has been WIP for a long time.
Please feel free to reopen if you want to continue.

**Pull request closed without merging**